src/python/pants/backend/project_info/depgraph_test.py

```diff
@@ -0,0 +1,177 @@
+            QueryRule(Addresses, (AddressSpecs,)),
+        ],
+        target_types=[PythonLibrary, PythonRequirementLibrary, SpecialDepsTarget],
+    )
+
+
+def test_depgraph_construction(rule_runner: RuleRunner) -> None:
+    rule_runner.write_files(
+        {
+            "3rdparty/python/BUILD": dedent(
+                """\
+            python_requirement_library(
+                name='extlib',
+                requirements=['extlib==1.2.3'],
+            )
+            """
+            ),
+            "src/python/foo/BUILD": dedent(
+                """\
+            python_library(dependencies=['3rdparty/python:extlib'])
+            """
+            ),
+            "src/python/foo/bar.py": "",
+            "src/python/baz/BUILD": dedent(
+                """\
+            python_library(name="lib", dependencies=['src/python/foo'])
+            """
+            ),
+            "src/python/baz/qux.py": "",
+        }
+    )
+
+    addresses = rule_runner.request(Addresses, [AddressSpecs([DescendantAddresses("")])])
+    dep_graph = rule_runner.request(
+        DependencyGraph, [DependencyGraphRequest(addresses=addresses, transitive=True)]
+    )
+    assert dep_graph == DependencyGraph(
+        vertices=[
+            Vertex(
+                "0",
+                {
+                    "address": "3rdparty/python:extlib",
+                    "type": "python_requirement_library",
+                    "requirements": ("extlib==1.2.3",),
+                },
+            ),
+            Vertex(
+                "1",
+                {
+                    "address": "src/python/baz:lib",
+                    "type": "python_library",
+                },
+            ),
+            Vertex(
+                "2",
+                {
+                    "address": "src/python/foo",
+                    "type": "python_library",
+                },
+            ),
+        ],
+        edges=[
+            Edge("1", "2", {}),
+            Edge("2", "0", {}),
+        ],
+    )
+
+    assert dep_graph.to_json() == dedent(
+        """\
+    {
+      "vertices": [
+        {
+          "id": "0",
+          "data": {
+            "address": "3rdparty/python:extlib",
+            "type": "python_requirement_library",
+            "requirements": [
+              "extlib==1.2.3"
+            ]
+          }
+        },
+        {
+          "id": "1",
+          "data": {
+            "address": "src/python/baz:lib",
+            "type": "python_library"
+          }
+        },
+        {
+          "id": "2",
+          "data": {
+            "address": "src/python/foo",
+            "type": "python_library"
+          }
+        }
+      ],
+      "edges": [
+        {
+          "src_id": "1",
+          "dep_id": "2",
+          "data": {}
+        },
+        {
+          "src_id": "2",
+          "dep_id": "0",
+          "data": {}
+        }
+      ]
+    }"""
+    )
+
+    assert dep_graph.to_dependencies_json() == dedent(
```

Comment:
This doesn't align with `dependees`, and is redundant with `peek`. I continue to think that keeping the output of this goal simple and composable would be to our users benefit, but I won't block.

Reply:
So... still not blocking, but: one more wildcard thought here (cc @Eric-Arellano). The desire to add more per-node information draw me back to comparison to `peek`

What if we:
1. added computation of the `dependencies` of a target to `peek` (...optionally?)
2. _did not add_ the `--output` mode to `dependencies`
2. _removed_ the `--output` mode from `dependees`

Conceptually, this would mean that the `dependencies` and `dependees` goals were always about rendering flat lists of addresses, and `peek` was always about rendering per-node information in the graph. Assuming the above changes, the way to get information about every node in the graph (including about its dependencies) would just be:
```console
./pants peek ::
```
To get information about only the dependencies of a particular target would be:
```
./pants dependencies --transitive $target | xargs ./pants peek
```

An open question is whether computing dependencies is too expensive for `peek` to do by default... i.e., whether you'd want a flag like `./pants peek --compute=dependencies ::` (leaving room for a `--compute=sources` for the other `AsyncField` type).

Reply:
This doesn't align with `dependees`... yet. As mentioned, my plan is to bring `dependees` into alignment with this.

I disagree that forcing users into complicated compositions (which, as discussed elsewhere, will cause pantsd to restart, with the attendant unnecessary and confusing performance hits) just to get a build graph seems unwieldy and not to their benefit.

`peek` seems like an odd name for something that returns a graph, but I'm not against it. It looks like peek already shows dependencies?

Reply:
> It looks like peek already shows dependencies?

It does, but it renders them verbatim from the BUILD file, rather than computing them via `DependenciesRequest`... hence the comment about optionally computing them. I expect that it would be fine to compute by default since it isn't transitive.

Reply:
I'm warming to the idea of `peek` being the only place with JSON output.

Reply:
Yea, me too... sorry I didn't think of it earlier =(

Reply:
Oh, interesting, so that would allow us to distinguish explicit and implicit deps?

Reply:
There's a bit of an impedance mismatch with `peek` as well. Unless we add a `--transitive` flag, there is no way to say "give me the dependency graph induced on this set of targets". `./pants peek ::` would work to get the entire graph, but subgraphs would not be easy to get. 

Your suggestion of `./pants dependencies --transitive $target | xargs ./pants peek` will omit $target itself from the output, which would be weird.

Reply:
That said, a `--transitive` flag on `peek` could make sense?

Reply:
I guess. This is still turning something that should be easy (I just want a dump of a dependency graph so I can run outside logic on it!) into a whole production (Run this thing with these magic flags, then pipe it into this other thing with a bunch of other magic flags that by the way will kill pantsd and thus your performance on any subsequent thing you run), but I'm tired of the debate at this point. The actual use case I care about is getting the full dependency graph, and at least that does not require pipes.

Reply:
And, not to belabor the point, I don't think we can tell people to compose pants runs as a reasonable solution to anything until we fix the pantsd issue there.

Reply:
People have been using pants runs as solutions for years before `pantsd` existed. So yes: we should fix that issue. But it's never been the case that there has been a single "do everything I want to happen in CI" goal.

Reply:
Yes, but with pantsd we have made that experience dramatically worse. It's not just some minor issue we should fix some day. If we're going to recommend that composing pants commands is a first class thing you should be doing, then we have to sort this out.

Reply:
#13028